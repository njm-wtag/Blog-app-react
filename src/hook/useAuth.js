import { useSelector } from "react-redux";

const useAuth = () => {
  const { authUser, success, error } = useSelector((state) => state.auth);

  return {
    authUser,
    success,
    error,
  };
};

export default useAuth;
