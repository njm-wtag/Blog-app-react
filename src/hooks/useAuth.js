import { useSelector } from "react-redux";

const useAuth = () => {
  const { authUser, success, error, errorMessage } = useSelector(
    (state) => state.auth
  );
  return {
    authUser,
    success,
    error,
    errorMessage,
  };
};

export default useAuth;
