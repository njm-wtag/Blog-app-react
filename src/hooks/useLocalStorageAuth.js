const useLocalStorageAuth = () => {
  const getAuthUser = localStorage.getItem("authUser");
  const auth = JSON.parse(getAuthUser);
  if (getAuthUser !== undefined && auth?.username && auth?.password) {
    return true;
  }
  return false;
};

export default useLocalStorageAuth;
