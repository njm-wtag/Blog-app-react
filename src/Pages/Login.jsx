import AuthForm from "components/AuthForm/AuthForm";
import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { loggedInUser } from "features/auth/authSlice";
import { useNavigate } from "react-router-dom";
import Layout from "components/Layout/Layout";
import useAuth from "../hook/useAuth";

const Login = () => {
  const { success } = useAuth();
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const handleSubmit = (values) => {
    dispatch(loggedInUser(values));
  };

  useEffect(() => {
    if (success) {
      navigate("/");
    }
  }, [success, navigate]);

  return (
    <Layout>
      <h1>Login</h1>

      <AuthForm handleSubmit={handleSubmit} />
    </Layout>
  );
};

export default Login;
