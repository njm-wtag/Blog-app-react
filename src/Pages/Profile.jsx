import { useEffect, useState } from "react";
import Layout from "components/Layout/Layout";
import { v4 as uuidv4 } from "uuid";
import AuthorDetails from "components/AuthorDetails/AuthorDetails";
import BlogList from "components/BlogList/BlogList";
import EditProfileForm from "components/EditProfileForm/EditProfileForm";
import { postBlog } from "features/blogs/blogsSlice";
import BlogForm from "components/BlogForm/BlogForm";
import useAuth from "hooks/useAuth";
import useBlogs from "hooks/useBlogs";
import ButtonContainer from "components/ButtonContainer/ButtonContainer";
import { useDispatch } from "react-redux";
import useSearch from "hooks/useSearch";

const Profile = () => {
  const [isEditProfileFormOpen, setIsEditProfileFormOpen] = useState(false);
  const [isAddBlogFormOpen, setIsAddBlogFormOpen] = useState(false);
  const { authUser } = useAuth();
  const blogs = useBlogs();
  const { queryInProfile } = useSearch();
  const dispatch = useDispatch();

  const blogByAuthor = blogs?.filter((blog) => blog.authorId === authUser.id);
  const onSubmit = (blog) => {
    blog.id = uuidv4();
    blog.authorId = authUser.id;
    blog.createdAt = new Date().toISOString();
    dispatch(postBlog(blog));
    setIsAddBlogFormOpen && setIsAddBlogFormOpen(false);
  };

  return (
    <Layout className="profile-page">
      <ButtonContainer
        setIsAddBlogFormOpen={setIsAddBlogFormOpen}
        setIsEditProfileFormOpen={setIsEditProfileFormOpen}
      />

      {authUser && <AuthorDetails authUser={authUser} />}
      {isEditProfileFormOpen && (
        <EditProfileForm setIsEditProfileFormOpen={setIsEditProfileFormOpen} />
      )}
      {isAddBlogFormOpen && (
        <BlogForm
          setIsAddBlogFormOpen={setIsAddBlogFormOpen}
          onSubmit={onSubmit}
        />
      )}
      <h3>My published posts</h3>
      {blogByAuthor.length ? (
        <BlogList blogs={blogByAuthor} query={queryInProfile} />
      ) : (
        "No blog published yet"
      )}
    </Layout>
  );
};

export default Profile;
