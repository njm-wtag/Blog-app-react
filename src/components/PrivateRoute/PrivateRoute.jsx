import { Navigate, Outlet } from "react-router-dom";
import useLocalStorageAuth from "../../hooks/useLocalStorageAuth";

const PrivateRoute = () => {
  const isUserloggedIn = useLocalStorageAuth();

  return isUserloggedIn ? <Outlet /> : <Navigate to={"/login"} />;
};

export default PrivateRoute;
