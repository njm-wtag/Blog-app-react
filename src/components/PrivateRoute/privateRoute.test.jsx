import { render, screen } from "@testing-library/react";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import PrivateRoute from ".";
import useLocalStorageAuth from "hooks/useLocalStorageAuth";
import Login from "pages/Login";
import Profile from "pages/Profile";
import useAuth from "hooks/useAuth";

vi.mock("hooks/useLocalStorageAuth");
vi.mock("hooks/useAuth");

describe("PrivateRoute Component", () => {
  it("navigates to login page when the user is not logged in", () => {
    useLocalStorageAuth.mockReturnValue(false);

    render(
      <BrowserRouter>
        <Routes>
          <Route path="/login" element={<Login />} />
        </Routes>
      </BrowserRouter>
    );
    const loginElement = screen.getByText(/Login/i);
    console.log(loginElement);
    // expect(loginElement).toBeInTheDocument();
  });

  // it("navigates to private route when the user is logged in", () => {
  //   useLocalStorageAuth.mockReturnValue(true);
  //   const mockAuthUser = {
  //     username: "John doe",
  //   };

  //   useAuth.mockReturnValue({ authUser: mockAuthUser });

  //   render(
  //     <BrowserRouter>
  //       <Routes>
  //         <Route element={<PrivateRoute />}>
  //           <Route path="/me" element={<Profile />} />
  //         </Route>
  //       </Routes>
  //     </BrowserRouter>
  //   );
  //   // console.log(container.firstChild);
  //   // expect(container.firstChild).toMatchSnapshot();
  //   const profileElement = screen.queryByText(mockAuthUser.username);
  //   console.log(profileElement);
  //   // expect(profileElement).toBeInTheDocument();
  // });
});
