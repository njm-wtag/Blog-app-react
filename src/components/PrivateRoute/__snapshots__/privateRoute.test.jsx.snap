// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrivateRoute Component > navigates to private route when the user is logged in 1`] = `null`;
